source/MaterialXTest/MaterialXGenOsl/GenOsl.cpp

```diff
@@ -102,6 +104,88 @@ TEST_CASE("GenShader: OSL Unique Names", "[genosl]")
+    //
+
+    mx::AttributeDefPtr adNodeName = doc->addAttributeDef("AD_node_name");
+    adNodeName->setType("string");
+    adNodeName->setAttrName("node_name");
+    adNodeName->setExportable(true);
+
+    mx::AttributeDefPtr adNodeCategory = doc->addAttributeDef("AD_node_category");
+    adNodeCategory->setType("string");
+    adNodeCategory->setAttrName("node_category");
+    adNodeCategory->setExportable(true);
+
+    mx::AttributeDefPtr adNodeTypeId = doc->addAttributeDef("AD_node_type_id");
+    adNodeTypeId->setType("integer");
+    adNodeTypeId->setAttrName("node_type_id");
+    adNodeTypeId->setExportable(true);
+
+    mx::AttributeDefPtr adAttributeLongName = doc->addAttributeDef("AD_attribute_long_name");
+    adAttributeLongName->setType("string");
+    adAttributeLongName->setAttrName("attribute_long_name");
+    adAttributeLongName->setExportable(true);
+
+    mx::AttributeDefPtr adAttributeShortName = doc->addAttributeDef("AD_attribute_short_name");
+    adAttributeShortName->setType("string");
+    adAttributeShortName->setAttrName("attribute_short_name");
+    adAttributeShortName->setExportable(true);
+
+    // Define a non-exportable attribute.
+    mx::AttributeDefPtr adShouldNotBeExported = doc->addAttributeDef("AD_should_not_be_exported");
+    adShouldNotBeExported->setType("float");
+    adShouldNotBeExported->setAttrName("should_not_be_exported");
+    adShouldNotBeExported->setExportable(false);
+
+    //
+    // Assign metadata on a shader nodedef
+    //
+
+    mx::NodeDefPtr stdSurfNodeDef = doc->getNodeDef("ND_standard_surface_surfaceshader");
+    REQUIRE(stdSurfNodeDef != nullptr);
+    stdSurfNodeDef->setAttribute("node_name", "StandardSurface");
+    stdSurfNodeDef->setAttribute("node_category", "shader/surface");
+    stdSurfNodeDef->setAttribute("node_type_id", "1234");
+
+    mx::InputPtr baseColor = stdSurfNodeDef->getInput("base_color");
+    REQUIRE(baseColor != nullptr);
+    baseColor->setAttribute("attribute_long_name", "BaseColor");
+    baseColor->setAttribute("attribute_short_name", "bc");
+    baseColor->setAttribute("should_not_be_exported", "42");
+
+    //
+    // Create an instance of this shader and validate that metadata is exported as expected.
+    //
+
+    mx::NodePtr stdSurf1 = doc->addNodeInstance(stdSurfNodeDef, "standardSurface1");
+    REQUIRE(stdSurf1 != nullptr);
+
+    mx::ShaderGeneratorPtr generator = mx::OslShaderGenerator::create();
+    mx::GenContext context(mx::OslShaderGenerator::create());
+    context.registerSourceCodeSearchPath(searchPath);
+
+    // Metadata to export must be registered in the context before shader generation starts.
+    // Custom generators can override this mehtod to customize which metadata gets registered.
+    generator->registerShaderMetadata(doc, context);
+
+    // Generate the shader and write to file for inspection.
+    mx::ShaderPtr shader = generator->generate(stdSurf1->getName(), stdSurf1, context);
```

Comment:
Query from before on if re-initialization occurs if you call register+generate() again but modify the meta data in-between.

Reply:
If called multiple times only new metadata entries will be registered, existing ones will not be overwritten. If you absolutely need to clear and reinitialize the registry you can do that by calling:

```
ShaderMetadataRegistryPtr registry = context.getUserData<ShaderMetadataRegistry>(ShaderMetadataRegistry::USER_DATA_NAME);
registry->clear();
```

But for most use-cases you would only call registerShaderMetadata once, at start up, when loading standard libraries. Similar to for example how color management is initialized.